File: skills/code-architect/references/code-examples/typescript/AbstractFactory.ts
// Abstract Factory — create families of related objects without specifying concrete classes.

interface Button {
  render(): string;
}

interface Switch {
  toggle(): string;
}

interface WidgetFactory {
  makeButton(): Button;
  makeSwitch(): Switch;
}

class LightButton implements Button {
  render(): string {
    return "[ Light Button ]";
  }
}

class LightSwitch implements Switch {
  toggle(): string {
    return "light switch flipped";
  }
}

class DarkButton implements Button {
  render(): string {
    return "[ Dark Button ]";
  }
}

class DarkSwitch implements Switch {
  toggle(): string {
    return "dark switch flipped";
  }
}

class LightThemeFactory implements WidgetFactory {
  makeButton(): Button {
    return new LightButton();
  }
  makeSwitch(): Switch {
    return new LightSwitch();
  }
}

class DarkThemeFactory implements WidgetFactory {
  makeButton(): Button {
    return new DarkButton();
  }
  makeSwitch(): Switch {
    return new DarkSwitch();
  }
}

function buildUi(factory: WidgetFactory): void {
  const button = factory.makeButton();
  const toggle = factory.makeSwitch();
  console.log(`${button.render()} -> ${toggle.toggle()}`);
}

function demo(): void {
  buildUi(new LightThemeFactory());
  buildUi(new DarkThemeFactory());
}

demo();

File: skills/code-architect/references/code-examples/swift/AbstractFactory.swift
// Abstract Factory — create families of related objects without specifying concrete classes.

protocol Button {
    func render() -> String
}

protocol Slider {
    func render() -> String
}

protocol WidgetFactory {
    func makeButton() -> Button
    func makeSlider() -> Slider
}

struct LightButton: Button {
    func render() -> String { "[ Light Button ]" }
}

struct LightSlider: Slider {
    func render() -> String { "—o— Light Slider" }
}

struct DarkButton: Button {
    func render() -> String { "█ Dark Button █" }
}

struct DarkSlider: Slider {
    func render() -> String { "■o■ Dark Slider" }
}

struct LightThemeFactory: WidgetFactory {
    func makeButton() -> Button { LightButton() }
    func makeSlider() -> Slider { LightSlider() }
}

struct DarkThemeFactory: WidgetFactory {
    func makeButton() -> Button { DarkButton() }
    func makeSlider() -> Slider { DarkSlider() }
}

func buildScreen(using factory: WidgetFactory) {
    print(factory.makeButton().render())
    print(factory.makeSlider().render())
}

func runDemo() {
    print("Light theme:")
    buildScreen(using: LightThemeFactory())
    print("Dark theme:")
    buildScreen(using: DarkThemeFactory())
}

runDemo()
